flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -730,6 +730,12 @@ private BufferOrEvent transformEvent(
 
         if (event.getClass() == EndOfPartitionEvent.class) {
             channelsWithEndOfPartitionEvents.set(currentChannel.getChannelIndex());
+            synchronized (inputChannelsWithData) {
+                enqueuedInputChannelsWithData.clear(currentChannel.getChannelIndex());
+                if (inputChannelsWithData.contains(currentChannel)) {
+                    inputChannelsWithData.getAndRemove(channel -> channel == currentChannel);
+                }
+            }
```

Comment:
I'd put it below `currentChannel.releaseAllResources()`:
- This method is called from task thread, so there is no risk in concurrent `getNextBuffer`.
- A concurrent `queueChannel` may come after this block and before `releaseAllResources`.

Reply:
> A concurrent queueChannel may come after this block and before releaseAllResources.

That's a very good point! 
I think it would be even better to check `channelsWithEndOfPartitionEvents` in `queueChannelUnsafe`.
WDYT? (I've updated the PR accordingly).